vm: move `ExpandToAst` logic into own instruction

zerbina opened the pull request:
The logic for template expansion as needed by the `mExpandToAst` magic
(used by `quote`) is now implemented as a separate instruction instead
of as part of `IndCallAsgn`. This makes the role of template expansion
less confusing ("why are templates 'called' inside the VM?", "weren't
they expanded already?") and also allows for a different instruction
argument make-up. The latter is required for both a future overhaul
of macro expansion and the refactored VM.

User-facing behaviour stays the same

saem commented on compiler/vm/vm.nim:
```diff
@@ -1287,6 +1287,32 @@ proc rawExecute(c: var TCtx, pc: var int, tos: var StackFrameIndex): TFullReg =
+    of opcExpandToAst:
+      decodeBC(rkNode)
+
+      let callExprAst = c.constants[regs[rb].intVal.int]
```
Please favour let blocks

saem commented on compiler/vm/vm.nim:
```diff
@@ -1287,6 +1287,32 @@ proc rawExecute(c: var TCtx, pc: var int, tos: var StackFrameIndex): TFullReg =
+        templCall.add(node)
+
+      var a = evalTemplate(templCall, prc, genSymOwner, c.config, c.cache, c.templInstCounter, c.idgen)
+      if a.kind == nkStmtList and a.len == 1: a = a[0]
```
```suggestion
      if a.kind == nkStmtList and a.len == 1: # flatten if a single statement
        a = a[0]
```

Also, reading this code, what if it's empty? Should this be an nkEmpty then?

zerbina replied:
I don't know. In order to preserve the previous behaviour, I left this as is

saem replied:
OK, let's stick with it for now, it does feel like a bug waiting to happen. 😬

saem reviewed (changes requested):
Make makes more sense that it works this way.

One question about the statement list handling, style bit if you're in there.

zerbina committed:
vm: move `ExpandToAst` logic into own instruction

The logic for template expansion as needed by the `mExpandToAst` magic
(used by `quote`) is now implemented as a separate instruction instead
of as part of `IndCallAsgn`. This makes the role of template expansion
less confusing ("why are templates 'called' inside the VM?", "weren't
they expanded already?") and also allows for a different instruction
argument make-up. The latter is required by both a future overhaul
of macro expansion and the refactored VM.

User-facing behaviour stays the same

saem reviewed (approved):
Bors r+
